lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +319,318 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
   QueuedTasks.push(std::move(T));
 }
 
-bool TaskQueue::execute(TaskBeganCallback Began, TaskFinishedCallback Finished,
-                        TaskSignalledCallback Signalled) {
+/// Owns Tasks, handles correspondence between Tasks, file descriptors, and
+/// process IDs.
+/// FIXME: only handles stdout pipes, ignores stderr pipes.
+class TaskMap {
   using PidToTaskMap = llvm::DenseMap<pid_t, std::unique_ptr<Task>>;
+  PidToTaskMap TasksByPid;
 
-  // Stores the current executing Tasks, organized by pid.
-  PidToTaskMap ExecutingTasks;
+public:
+  TaskMap() = default;
 
-  // Maintains the current fds we're checking with poll.
-  std::vector<struct pollfd> PollFds;
+  bool empty() const { return TasksByPid.empty(); }
+  unsigned size() const { return TasksByPid.size(); }
 
-  bool SubtaskFailed = false;
+  void add(std::unique_ptr<Task> T) { TasksByPid[T->getPid()] = std::move(T); }
 
-  unsigned MaxNumberOfParallelTasks = getNumberOfParallelTasks();
+  Task &findTaskForFd(const int fd) {
+    auto predicate = [&fd](PidToTaskMap::value_type &value) -> bool {
+      return value.second->getPipe() == fd;
+    };
+    auto iter = std::find_if(TasksByPid.begin(), TasksByPid.end(), predicate);
+    assert(iter != TasksByPid.end() &&
+           "All outstanding fds must be associated with a  Task");
+    return *iter->second;
+  }
 
-  if (MaxNumberOfParallelTasks == 0)
-    MaxNumberOfParallelTasks = 1;
+  void destroyTask(Task &T) { TasksByPid.erase(T.getPid()); }
+};
 
-  while ((!QueuedTasks.empty() && !SubtaskFailed) ||
-         !ExecutingTasks.empty()) {
-    // Enqueue additional tasks, if we have additional tasks, we aren't
-    // already at the parallel limit, and no earlier subtasks have failed.
-    while (!SubtaskFailed && !QueuedTasks.empty() &&
-           ExecutingTasks.size() < MaxNumberOfParallelTasks) {
-      std::unique_ptr<Task> T(QueuedTasks.front().release());
-      QueuedTasks.pop();
-      if (T->execute())
-        return true;
+/// Concurrently execute the tasks in the TaskQueue, collecting the outputs from
+/// each task. The typical task is a Swift frontend job.
```

Comment:
> The typical task is a Swift frontend job.

Remove sentence I think (especially if this is to-be-extracted to LLVM later).

Reply:
Done!